### PR TITLE
feat(stream): `source_error_count` reporting to prometheus 

### PR DESCRIPTION
**Pull request by jon-chuang**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?
Source stream error reporting to prometheus 

image

## Checklist

- [x] I have written necessary rustdoc comments
- [x] I have added necessary unit tests and integration tests
- [x] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features).
- [x] I have demonstrated that backward compatibility is not broken by breaking changes and created issues to track deprecated features to be removed in the future. (Please refer to the issue)
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)

# Issue
https://github.com/risingwavelabs/risingwave/issues/7831

### CODE REVIEW

File: src/connector/src/source/base.rs [hunk 1]
```diff
@@ -80,13 +80,78 @@ pub trait SplitEnumerator: Sized {
...
+    pub fn new(actor_id: u32, source_id: TableId, fragment_id: u32) -> Self {
         SourceInfo {
             actor_id,
             source_id,
+            fragment_id,
+        }
+    }
+}
+
+// const MAX_BLACKLISTED_STREAM_ERROR_MSGS: usize = 50;
```
Comment (fuyufjh) [inline src/connector/src/source/base.rs:97]: Obsolete code?

File: src/connector/src/source/base.rs [hunk 2]
```diff
@@ -80,13 +80,78 @@ pub trait SplitEnumerator: Sized {
...
+//     fn should_truncate_msg(&self, msg: &str) -> bool {
+//         if !self.past_error_msgs.read().contains(msg) {
+
+//         }
+//         false
+//     }
+// }
+
+#[derive(Debug, Clone)]
+pub struct ErrorReportingContext {
```
Comment (fuyufjh) [inline src/connector/src/source/base.rs:116]: Perhaps can be `Context` or `SourceContext` to make it shorter. Also implies it's okay to add more things here in the future.

File: src/connector/src/parser/json_parser.rs [hunk 3]
```diff
@@ -20,19 +20,33 @@ use simd_json::{BorrowedValue, ValueAccess};
...
-use crate::source::SourceColumnDesc;
+use crate::source::{ErrorReportingContext, SourceColumnDesc};
 
 impl_common_parser_logic!(JsonParser);
 
 /// Parser for JSON format
 #[derive(Debug)]
 pub struct JsonParser {
     rw_columns: Vec<SourceColumnDesc>,
+    error_ctx: ErrorReportingContext,
```
Comment (fuyufjh) [inline src/connector/src/parser/json_parser.rs:31]: It is not actually used in `JsonParser`?

Comment (fuyufjh) [reply src/connector/src/parser/json_parser.rs:31]: Same problem for all the other Parsers.

File: src/frontend/src/session.rs [hunk 4]
```diff
@@ -12,11 +12,11 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
+// use tokio::sync::Mutex;
```
Comment (fuyufjh) [inline src/frontend/src/session.rs:15]:
```suggestion
```

File: src/stream/src/executor/monitor/streaming_stats.rs [hunk 5]
```diff
@@ -433,9 +433,9 @@ impl StreamingMetrics {
         )
         .unwrap();
 
-        let user_error_count = register_int_counter_vec_with_registry!(
-            "user_error_count",
-            "user errors in the system, queryable by tags",
+        let compute_error_count = register_int_counter_vec_with_registry!(
+            "compute_error_count",
```
Comment (fuyufjh) [inline src/stream/src/executor/monitor/streaming_stats.rs:437]: What is this change for? Does it mean that non-user errors will be included?

File: src/connector/src/source/base.rs (hunk 2 above)
Comment (tabVersion) [reply src/connector/src/source/base.rs:116]: prefer `SourceErrorConext`

File: src/connector/src/source/monitor/metrics.rs [hunk 6]
```diff
@@ -38,10 +40,24 @@ impl SourceMetrics {
...
+        let source_error_count = register_int_counter_vec_with_registry!(
+            "source_error_count",
+            "Source errors in the system, queryable by tags",
+            &[
+                "error_type",
+                "error_msg",
+                "executor_name",
+                "fragment_id",
+                "table_id"
+            ],
```
Comment (tabVersion) [inline src/connector/src/source/monitor/metrics.rs:52]: why do we need fragment id here?
the same source belonging to different fragments will have the same errors.

**Review by tabVersion — COMMENTED** [review]
basically LGTM, but need some code cleaning

File: src/connector/src/parser/json_parser.rs (hunk 3 above)
Comment (jon-chuang) [reply src/connector/src/parser/json_parser.rs:31]: It is by `impl_common_parser_logic!`

File: src/stream/src/executor/monitor/streaming_stats.rs (hunk 5 above)
Comment (fuyufjh) [reply src/stream/src/executor/monitor/streaming_stats.rs:437]: +1 for keeping `user` there

**Review by fuyufjh — APPROVED** [review]
LGTM

**Pull request merged**